gensim/models/ensemblelda.py

```diff
@@ -0,0 +1,1315 @@
+
+        # _COSINE_DISTANCE_CALCULATION_THRESHOLD is used so that cosine distance calculations can be sped up by skipping
+        # distance calculations for highly masked topic-term distributions
+        self._COSINE_DISTANCE_CALCULATION_THRESHOLD = 0.05
+
+        if "id2word" not in gensim_kw_args:
+            gensim_kw_args["id2word"] = None
+        if "corpus" not in gensim_kw_args:
+            gensim_kw_args["corpus"] = None
+
+        if gensim_kw_args["id2word"] is None and not gensim_kw_args["corpus"] is None:
+            logger.warning("no word id mapping provided; initializing from corpus, assuming identity")
+            gensim_kw_args["id2word"] = utils.dict_from_corpus(gensim_kw_args["corpus"])
+        if gensim_kw_args["id2word"] is None and gensim_kw_args["corpus"] is None:
+            raise ValueError(
+                "at least one of corpus/id2word must be specified, to establish "
+                "input space dimensionality. Corpus should be provided using the "
+                "`corpus` keyword argument."
+            )
+
+        if type(topic_model_class) == type and issubclass(topic_model_class, ldamodel.LdaModel):
+            self.topic_model_class = topic_model_class
+        else:
+            kinds = {
+                "lda": ldamodel.LdaModel,
+                "ldamulticore": ldamulticore.LdaMulticore
+            }
+            if topic_model_class not in kinds:
+                raise ValueError(
+                    "topic_model_class should be one of 'lda', 'ldamulticode' or a model "
+                    "inheriting from LdaModel"
+                )
+            self.topic_model_class = kinds[topic_model_class]
+
+        self.num_models = num_models
+        self.gensim_kw_args = gensim_kw_args
+
+        self.memory_friendly_ttda = memory_friendly_ttda
+
+        self.distance_workers = distance_workers
+        self.masking_threshold = masking_threshold
+        self.masking_method = masking_method
+
+        # this will provide the gensim api to the ensemble basically
+        self.classic_model_representation = None
+
+        # the ensembles state
+        self.random_state = utils.get_random_state(random_state)
+        self.sstats_sum = 0
+        self.eta = None
+        self.tms = []
+        # initialize empty topic term distribution array
+        self.ttda = np.empty((0, len(gensim_kw_args["id2word"])))
+        self.asymmetric_distance_matrix_outdated = True
+
+        # in case the model will not train due to some
+        # parameters, stop here and don't train.
+        if num_models <= 0:
+            return
+        if gensim_kw_args.get("corpus") is None:
+            return
+        if "iterations" in gensim_kw_args and gensim_kw_args["iterations"] <= 0:
+            return
+        if "passes" in gensim_kw_args and gensim_kw_args["passes"] <= 0:
+            return
+
+        logger.info("generating {} topic models...".format(num_models))
```

Comment:
Use f-string formatting please. We avoid `.format` in Gensim.

(here and everywhere)

Reply:
done